graphql_compiler/query_pagination/query_parameterizer.py

```diff
@@ -1,61 +1,194 @@
+            if field_name == query_path[0]:
+                found_field = True
+                new_selection_ast, sub_removed_parameters = _add_pagination_filter(
+                    selection_ast, query_path[1:], pagination_field, directive_to_add
+                )
+                removed_parameters.extend(sub_removed_parameters)
+            new_selections.append(new_selection_ast)
+
+        if not found_field:
+            raise AssertionError()
+
+    new_ast = copy(query_ast)
+    new_ast.selection_set = SelectionSetNode(selections=new_selections)
+    return new_ast, removed_parameters
+
+
+def _make_binary_filter_directive_node(op_name: str, param_name: str) -> DirectiveNode:
+    """Make a binary filter directive node with the given binary op_name and parameter name."""
+    return DirectiveNode(
+        name=NameNode(value="filter"),
+        arguments=[
+            ArgumentNode(name=NameNode(value="op_name"), value=StringValueNode(value=op_name),),
+            ArgumentNode(
+                name=NameNode(value="value"),
+                value=ListValueNode(values=[StringValueNode(value="$" + param_name)]),
+            ),
+        ],
+    )
+
+
+def generate_parameterized_queries(
+    schema_info: QueryPlanningSchemaInfo,
+    query_ast: DocumentNode,
+    parameters: Dict[str, Any],
+    vertex_partition: VertexPartitionPlan,
+) -> Tuple[Tuple[DocumentNode, Dict[str, Any]], Tuple[DocumentNode, Dict[str, Any]], str]:
+    """Generate two parameterized queries that can be used to paginate over a given query.
 
     Args:
         schema_info: QueryPlanningSchemaInfo
         query_ast: Document, query that is being paginated.
         parameters: dict, list of parameters for the given query.
+        vertex_partition: pagination plan
 
     Returns:
-        ParameterizedPaginationQueries namedtuple
+        # TODO use ASTWithParameters struct to simplify API
+        next_page: Ast and params for next page.
+        remainder_ast: Ast and params for remainder.
+        param_name: The parameter name used in the new filters.
     """
-    raise NotImplementedError()
+    query_type = get_only_query_definition(query_ast, GraphQLError)
+
+    param_name = _generate_new_name("__paged_param", set(parameters.keys()))
+    next_page_root, next_page_removed_parameters = _add_pagination_filter(
+        query_type,
+        vertex_partition.query_path,
+        vertex_partition.pagination_field,
+        _make_binary_filter_directive_node("<", param_name),
+    )
+    remainder_root, remainder_removed_parameters = _add_pagination_filter(
+        query_type,
+        vertex_partition.query_path,
+        vertex_partition.pagination_field,
+        _make_binary_filter_directive_node(">=", param_name),
+    )
+
+    page_parameters = {k: v for k, v in parameters.items() if k not in next_page_removed_parameters}
+    remainder_parameters = {
+        k: v for k, v in parameters.items() if k not in remainder_removed_parameters
+    }
```

Comment:
Here we remove the parameters corresponding to any redundant filters that were replaced with stronger filters. The new filters should be stronger since the parameter generator takes the existing filters into account, and makes sure to generate values in between the existing range of possible values. But still, it would be nice to assert that this new filter is indeed stronger. However, we can't do that because we don't have the values to be inserted yet.

@obi1kenobi This is the biggest doubt I have about the current API. Let me know if you see a better way to guarantee correctness.